File: src/com/meti/compile/lex/parse/structure/StructDeclareFactory.java
package com.meti.compile.lex.parse.structure;

import com.meti.compile.lex.Tokenizer;
import com.meti.compile.lex.parse.TokenizerFactory;
import com.meti.compile.node.Token;
import com.meti.compile.node.ValueToken;
import com.meti.compile.node.structure.StructDeclareToken.Builder;
import com.meti.util.*;

import java.util.List;

public class StructDeclareFactory implements TokenizerFactory {
    @Override
    public Tokenizer create(String content) {
        return new TokenizerImpl(content);
    }

    private static class TokenizerImpl implements Tokenizer {
        private final String content;

        public TokenizerImpl(String content) {
            this.content = content;
        }

        private Token parseFields(Builder builder) {
            return extractChildren()
                    .filter(s -> !s.isBlank())
                    .map(String::trim)
                    .map(ValueToken::new)
                    .reduce(builder, Builder::withChild)
                    .build();
        }

        private MonadStream<String> extractChildren() {
            return Monad.Monad(content)
                    .extract(value -> value.indexOf('{'))
                    .extractStart(value -> value.indexOf('}'))
                    .map((value, start, end) -> value.substring(start + 1, end))
                    .map(String::trim)
                    .map(value -> value.split(","))
                    .map(List::of)
                    .apply(MonadStream::Stream);
        }

        private Builder createBuilder() {
            return Monad.Monad(content)
                    .extract(value -> value.indexOf('<'))
                    .extractStart(value -> value.indexOf('>'))
                    .map((value, start, end) -> value.substring(start + 1, end))
                    .map(String::trim)
                    .with(new Builder())
                    .reverse()
                    .apply(Builder::withName);
        }

        @Override
        public Option<Token> evaluate() {
            return Some.Some(content)
                    .filter(value -> value.startsWith("<"))
                    .filter(value -> value.contains(">"))
                    .filter(value -> value.contains("{"))
                    .filter(value -> value.endsWith("}"))
                    .replace(this::createBuilder)
                    .map(this::parseFields);
        }
    }
}

File: src/com/meti/compile/lex/parse/structure/StructDeclareFactory.scala
package com.meti.compile.lex.parse.structure

import com.meti.compile.lex.Tokenizer
import com.meti.compile.lex.parse.TokenizerFactory
import com.meti.compile.node.structure.StructDeclareToken.Builder
import com.meti.compile.node.{Token, ValueToken}
import com.meti.util
import com.meti.util.Monad.Monad
import com.meti.util.MonadStream._
import com.meti.util.Some.Some

class TokenizerImpl(val content: String) extends Tokenizer {
  override def evaluate: util.Option[Token] = Some(content)
    .filter((value: String) => value.startsWith("<"))
    .filter((value: String) => value.contains(">"))
    .filter((value: String) => value.contains("{"))
    .filter((value: String) => value.endsWith("}"))
    .supplyValue(createBuilder)
    .map(parseFields)

  private def parseFields(builder: Builder) = extractChildren
    .filter((child: String) => !child.isEmpty)
    .map(_.trim)
    .map((value: String) => new ValueToken(value))
    .reduce(builder, (_: Builder).withChild(_: Token))
    .build()

  private def extractChildren = Monad(content)
    .extract((value: String) => value.indexOf('{'))
    .extractStart(_.indexOf('}'))
    .map((value: String, start: Int, `end`: Int) => value.substring(start + 1, `end`))
    .map(_.trim)
    .map(_.split(","))
    .apply(StreamArray(_))

  def createBuilder: Builder = Monad(content)
    .extract((value: String) => value.indexOf('<'))
    .extractStart((value: String) => value.indexOf('>'))
    .map((value: String, start: Int, end: Int) => value.substring(start + 1, end))
    .map(_.trim)
    .append(new Builder())
    .reverse()
    .apply((_: Builder).withName(_: String));
}

class StructDeclareFactory extends TokenizerFactory {
  override def create(content: String) = new TokenizerImpl(content)
}